File: 06 - LOGICIEL EMBARQUE/simulateur/src/config-store.js
const MAGIC = [0x56, 0x43, 0x4f, 0x32]; // "VCO2"
export const CONFIG_SCHEMA_VERSION = 1;
export const CONFIG_RECORD_SIZE = 32;
const PAYLOAD_SIZE = 20;

export const FACTORY_CONFIG = Object.freeze({
  targetPpm: 1000,
  minOutputPct: 20,
  maxOutputPct: 100,
  highAlarmPpm: 1500,
  alarmDelaySeconds: 60,
  testDurationSeconds: 600,
  historyPeriodSeconds: 60,
});

const crcTable = Array.from({ length: 256 }, (_, index) => {
  let value = index;
  for (let bit = 0; bit < 8; bit += 1) value = (value & 1) ? (0xedb88320 ^ (value >>> 1)) : (value >>> 1);
  return value >>> 0;
});

export function crc32(bytes) {
  let crc = 0xffffffff;
  for (const byte of bytes) crc = crcTable[(crc ^ byte) & 0xff] ^ (crc >>> 8);
  return (crc ^ 0xffffffff) >>> 0;
}

export function validatePersistentConfig(config) {
  const errors = [];
  const expectedKeys = ["alarmDelaySeconds", "highAlarmPpm", "historyPeriodSeconds", "maxOutputPct", "minOutputPct", "targetPpm", "testDurationSeconds"];
  const actualKeys = config && typeof config === "object" ? Object.keys(config).sort() : [];
  if (JSON.stringify(actualKeys) !== JSON.stringify(expectedKeys)) errors.push("schemaKeys");
  const integerInRange = (value, min, max) => Number.isInteger(value) && value >= min && value <= max;
  const tenthInRange = (value, min, max) => Number.isFinite(value) && value >= min && value <= max
    && Math.abs(value * 10 - Math.round(value * 10)) < 1e-9;
  if (!integerInRange(config?.targetPpm, 800, 1400)) errors.push("targetPpm");
  if (!tenthInRange(config?.minOutputPct, 0, 80)) errors.push("minOutputPct");
  if (!tenthInRange(config?.maxOutputPct, 50, 100)) errors.push("maxOutputPct");
  if (!integerInRange(config?.highAlarmPpm, 1200, 2500)) errors.push("highAlarmPpm");
  if (!integerInRange(config?.alarmDelaySeconds, 10, 600)) errors.push("alarmDelaySeconds");
  if (!integerInRange(config?.testDurationSeconds, 30, 600)) errors.push("testDurationSeconds");
  if (!integerInRange(config?.historyPeriodSeconds, 10, 300)) errors.push("historyPeriodSeconds");
  if (Number.isFinite(config?.minOutputPct) && Number.isFinite(config?.maxOutputPct)
    && config.minOutputPct > config.maxOutputPct) errors.push("minGreaterThanMax");
  if (Number.isFinite(config?.targetPpm) && Number.isFinite(config?.highAlarmPpm)
    && config.highAlarmPpm <= config.targetPpm) errors.push("alarmNotAboveTarget");
  return { ok: errors.length === 0, errors };
}

export function encodeConfigRecord(config, generation) {
  const validation = validatePersistentConfig(config);
  if (!validation.ok) throw new RangeError(`Configuration persistante invalide : ${validation.errors.join(", ")}`);
  if (!Number.isInteger(generation) || generation < 0 || generation > 0xffffffff) {
    throw new RangeError("Génération invalide.");
  }
  const bytes = new Uint8Array(CONFIG_RECORD_SIZE);
  bytes.set(MAGIC, 0);
  const view = new DataView(bytes.buffer);
  view.setUint16(4, CONFIG_SCHEMA_VERSION, true);
  view.setUint16(6, PAYLOAD_SIZE, true);
  view.setUint32(8, generation >>> 0, true);
  view.setUint16(12, config.targetPpm, true);
  view.setUint16(14, Math.round(config.minOutputPct * 10), true);
  view.setUint16(16, Math.round(config.maxOutputPct * 10), true);
  view.setUint16(18, config.highAlarmPpm, true);
  view.setUint16(20, config.alarmDelaySeconds, true);
  view.setUint16(22, config.testDurationSeconds, true);
  view.setUint16(24, config.historyPeriodSeconds, true);
  view.setUint16(26, 0, true);
  view.setUint32(28, crc32(bytes.subarray(0, 28)), true);
  return bytes;
}

export function decodeConfigRecord(input) {
  if (input == null) return { ok: false, reason: "EMPTY" };
  const bytes = input instanceof Uint8Array ? input : new Uint8Array(input);
  if (bytes.byteLength !== CONFIG_RECORD_SIZE) return { ok: false, reason: "LENGTH" };
  if (!MAGIC.every((value, index) => bytes[index] === value)) return { ok: false, reason: "MAGIC" };
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  if (view.getUint16(4, true) !== CONFIG_SCHEMA_VERSION) return { ok: false, reason: "SCHEMA" };
  if (view.getUint16(6, true) !== PAYLOAD_SIZE) return { ok: false, reason: "PAYLOAD_LENGTH" };
  if (view.getUint16(26, true) !== 0) return { ok: false, reason: "FLAGS" };
  if (view.getUint32(28, true) !== crc32(bytes.subarray(0, 28))) return { ok: false, reason: "CRC" };
  const config = {
    targetPpm: view.getUint16(12, true),
    minOutputPct: view.getUint16(14, true) / 10,
    maxOutputPct: view.getUint16(16, true) / 10,
    highAlarmPpm: view.getUint16(18, true),
    alarmDelaySeconds: view.getUint16(20, true),
    testDurationSeconds: view.getUint16(22, true),
    historyPeriodSeconds: view.getUint16(24, true),
  };
  const validation = validatePersistentConfig(config);
  if (!validation.ok) return { ok: false, reason: "BOUNDS", errors: validation.errors };
  return { ok: true, generation: view.getUint32(8, true), config };
}

export function isNewerGeneration(candidate, reference) {
  const difference = (candidate - reference) >>> 0;
  return difference !== 0 && difference < 0x80000000;
}

const sameConfig = (left, right) => JSON.stringify(left) === JSON.stringify(right);

export function loadConfigSlots(slots) {
  const a = decodeConfigRecord(slots?.a);
  const b = decodeConfigRecord(slots?.b);
  if (!a.ok && !b.ok) {
    const virgin = a.reason === "EMPTY" && b.reason === "EMPTY";
    return {
      config: { ...FACTORY_CONFIG },
      generation: 0,
      source: "FACTORY",
      status: virgin ? "FACTORY_DEFAULTS" : "CONFIG_CORRUPT",
      safeToAuto: virgin,
      slots: { a, b },
    };
  }
  if (a.ok && b.ok) {
    if (a.generation === b.generation) {
      if (!sameConfig(a.config, b.config)) {
        return {
          config: { ...FACTORY_CONFIG }, generation: a.generation, source: "FACTORY",
          status: "GENERATION_CONFLICT", safeToAuto: false, slots: { a, b },
        };
      }
      return { config: a.config, generation: a.generation, source: "A", status: "REDUNDANT", safeToAuto: true, slots: { a, b } };
    }
    const aNewer = isNewerGeneration(a.generation, b.generation);
    const bNewer = isNewerGeneration(b.generation, a.generation);
    if (aNewer === bNewer) {
      return {
        config: { ...FACTORY_CONFIG }, generation: 0, source: "FACTORY",
        status: "GENERATION_AMBIGUOUS", safeToAuto: false, slots: { a, b },
      };
    }
    const selected = aNewer ? a : b;
    return { config: selected.config, generation: selected.generation, source: aNewer ? "A" : "B", status: "OK", safeToAuto: true, slots: { a, b } };
  }
  const selected = a.ok ? a : b;
  return {
    config: selected.config,
    generation: selected.generation,
    source: a.ok ? "A" : "B",
    status: "BACKUP_DEGRADED",
    safeToAuto: true,
    slots: { a, b },
  };
}

export function planConfigWrite(slots, config, { allowRecovery = false } = {}) {
  const validation = validatePersistentConfig(config);
  if (!validation.ok) return { ok: false, reason: "INVALID_CONFIG", errors: validation.errors };
  const current = loadConfigSlots(slots);
  if (!current.safeToAuto && current.status !== "FACTORY_DEFAULTS" && !allowRecovery) {
    return { ok: false, reason: "RECOVERY_CONFIRMATION_REQUIRED", status: current.status };
  }
  if (["OK", "REDUNDANT"].includes(current.status) && sameConfig(current.config, config)) {
    return { ok: true, noOp: true, generation: current.generation, targetSlot: current.source.toLowerCase() };
  }
  const generation = (current.generation + 1) >>> 0;
  let targetSlot = "a";
  if (current.source === "A") targetSlot = "b";
  else if (current.source === "B") targetSlot = "a";
  return { ok: true, noOp: false, generation, targetSlot, record: encodeConfigRecord(config, generation) };
}

File: 06 - LOGICIEL EMBARQUE/firmware-esp32/components/config_store/config_store.cpp
#include "config_store.hpp"

#include <algorithm>
#include <cmath>
#include <cstring>

namespace ventilation {
namespace {

uint16_t read_u16_le(const uint8_t* data, size_t offset) {
  return static_cast<uint16_t>(data[offset] | (static_cast<uint16_t>(data[offset + 1U]) << 8U));
}

uint32_t read_u32_le(const uint8_t* data, size_t offset) {
  return static_cast<uint32_t>(data[offset]) |
         (static_cast<uint32_t>(data[offset + 1U]) << 8U) |
         (static_cast<uint32_t>(data[offset + 2U]) << 16U) |
         (static_cast<uint32_t>(data[offset + 3U]) << 24U);
}

void write_u16_le(std::array<uint8_t, kConfigRecordSize>& data, size_t offset, uint16_t value) {
  data[offset] = static_cast<uint8_t>(value & 0xffU);
  data[offset + 1U] = static_cast<uint8_t>((value >> 8U) & 0xffU);
}

void write_u32_le(std::array<uint8_t, kConfigRecordSize>& data, size_t offset, uint32_t value) {
  data[offset] = static_cast<uint8_t>(value & 0xffU);
  data[offset + 1U] = static_cast<uint8_t>((value >> 8U) & 0xffU);
  data[offset + 2U] = static_cast<uint8_t>((value >> 16U) & 0xffU);
  data[offset + 3U] = static_cast<uint8_t>((value >> 24U) & 0xffU);
}

bool tenth_in_range(float value, float min, float max) {
  if (!std::isfinite(value) || value < min || value > max) return false;
  return std::fabs(value * 10.0F - std::round(value * 10.0F)) < 0.0001F;
}

bool same_config(const PersistentConfig& left, const PersistentConfig& right) {
  return left.target_ppm == right.target_ppm &&
         std::fabs(left.min_output_pct - right.min_output_pct) < 0.0001F &&
         std::fabs(left.max_output_pct - right.max_output_pct) < 0.0001F &&
         left.high_alarm_ppm == right.high_alarm_ppm &&
         left.alarm_delay_s == right.alarm_delay_s &&
         left.test_duration_s == right.test_duration_s &&
         left.history_period_s == right.history_period_s;
}

ConfigDecodeResult decode_empty(ConfigDecodeReason reason) {
  return {false, reason, 0U, factory_config()};
}

ConfigLoadResult load_from_single_valid(const ConfigDecodeResult& selected,
                                        const ConfigDecodeResult& a,
                                        const ConfigDecodeResult& b) {
  const bool from_a = a.ok;
  return {
      selected.config,
      selected.generation,
      from_a ? ConfigSource::kSlotA : ConfigSource::kSlotB,
      ConfigLoadStatus::kBackupDegraded,
      true,
      a,
      b,
  };
}

}  // namespace

PersistentConfig factory_config() {
  return {};
}

bool validate_persistent_config(const PersistentConfig& config) {
  if (config.target_ppm < 800U || config.target_ppm > 1400U) return false;
  if (!tenth_in_range(config.min_output_pct, 0.0F, 80.0F)) return false;
  if (!tenth_in_range(config.max_output_pct, 50.0F, 100.0F)) return false;
  if (config.min_output_pct > config.max_output_pct) return false;
  if (config.high_alarm_ppm < 1200U || config.high_alarm_ppm > 2500U) return false;
  if (config.high_alarm_ppm <= config.target_ppm) return false;
  if (config.alarm_delay_s < 10U || config.alarm_delay_s > 600U) return false;
  if (config.test_duration_s < 30U || config.test_duration_s > 600U) return false;
  if (config.history_period_s < 10U || config.history_period_s > 300U) return false;
  return true;
}

uint32_t crc32_config(const uint8_t* data, size_t length) {
  uint32_t crc = 0xffffffffU;
  for (size_t index = 0; index < length; ++index) {
    crc ^= data[index];
    for (uint8_t bit = 0U; bit < 8U; ++bit) {
      crc = (crc & 1U) ? (0xedb88320U ^ (crc >> 1U)) : (crc >> 1U);
    }
  }
  return crc ^ 0xffffffffU;
}

bool encode_config_record(const PersistentConfig& config, uint32_t generation, ConfigRecord& output) {
  if (!validate_persistent_config(config)) return false;
  output.bytes.fill(0U);
  std::copy(kConfigMagic.begin(), kConfigMagic.end(), output.bytes.begin());
  write_u16_le(output.bytes, 4U, kConfigSchemaVersion);
  write_u16_le(output.bytes, 6U, static_cast<uint16_t>(kConfigPayloadSize));
  write_u32_le(output.bytes, 8U, generation);
  write_u16_le(output.bytes, 12U, config.target_ppm);
  write_u16_le(output.bytes, 14U, static_cast<uint16_t>(std::lround(config.min_output_pct * 10.0F)));
  write_u16_le(output.bytes, 16U, static_cast<uint16_t>(std::lround(config.max_output_pct * 10.0F)));
  write_u16_le(output.bytes, 18U, config.high_alarm_ppm);
  write_u16_le(output.bytes, 20U, config.alarm_delay_s);
  write_u16_le(output.bytes, 22U, config.test_duration_s);
  write_u16_le(output.bytes, 24U, config.history_period_s);
  write_u16_le(output.bytes, 26U, 0U);
  write_u32_le(output.bytes, 28U, crc32_config(output.bytes.data(), 28U));
  return true;
}

ConfigDecodeResult decode_config_record(const uint8_t* data, size_t length) {
  if (data == nullptr) return decode_empty(ConfigDecodeReason::kEmpty);
  if (length != kConfigRecordSize) return decode_empty(ConfigDecodeReason::kLength);
  if (!std::equal(kConfigMagic.begin(), kConfigMagic.end(), data)) return decode_empty(ConfigDecodeReason::kMagic);
  if (read_u16_le(data, 4U) != kConfigSchemaVersion) return decode_empty(ConfigDecodeReason::kSchema);
  if (read_u16_le(data, 6U) != kConfigPayloadSize) return decode_empty(ConfigDecodeReason::kPayloadLength);
  if (read_u16_le(data, 26U) != 0U) return decode_empty(ConfigDecodeReason::kFlags);
  if (read_u32_le(data, 28U) != crc32_config(data, 28U)) return decode_empty(ConfigDecodeReason::kCrc);

  PersistentConfig config{
      read_u16_le(data, 12U),
      static_cast<float>(read_u16_le(data, 14U)) / 10.0F,
      static_cast<float>(read_u16_le(data, 16U)) / 10.0F,
      read_u16_le(data, 18U),
      read_u16_le(data, 20U),
      read_u16_le(data, 22U),
      read_u16_le(data, 24U),
  };
  if (!validate_persistent_config(config)) return decode_empty(ConfigDecodeReason::kBounds);
  return {true, ConfigDecodeReason::kOk, read_u32_le(data, 8U), config};
}

bool is_newer_generation(uint32_t candidate, uint32_t reference) {
  const uint32_t difference = candidate - reference;
  return difference != 0U && difference < 0x80000000U;
}

ConfigLoadResult load_config_slots(const uint8_t* slot_a, size_t slot_a_length,
                                   const uint8_t* slot_b, size_t slot_b_length) {
  const ConfigDecodeResult a = decode_config_record(slot_a, slot_a_length);
  const ConfigDecodeResult b = decode_config_record(slot_b, slot_b_length);

  if (!a.ok && !b.ok) {
    const bool virgin = a.reason == ConfigDecodeReason::kEmpty && b.reason == ConfigDecodeReason::kEmpty;
    return {factory_config(), 0U, ConfigSource::kFactory,
            virgin ? ConfigLoadStatus::kFactoryDefaults : ConfigLoadStatus::kConfigCorrupt,
            virgin, a, b};
  }

  if (a.ok && b.ok) {
    if (a.generation == b.generation) {
      if (!same_config(a.config, b.config)) {
        return {factory_config(), a.generation, ConfigSource::kFactory,
                ConfigLoadStatus::kGenerationConflict, false, a, b};
      }
      return {a.config, a.generation, ConfigSource::kSlotA, ConfigLoadStatus::kRedundant, true, a, b};
    }
    const bool a_newer = is_newer_generation(a.generation, b.generation);
    const bool b_newer = is_newer_generation(b.generation, a.generation);
    if (a_newer == b_newer) {
      return {factory_config(), 0U, ConfigSource::kFactory,
              ConfigLoadStatus::kGenerationAmbiguous, false, a, b};
    }
    const ConfigDecodeResult& selected = a_newer ? a : b;
    return {selected.config, selected.generation, a_newer ? ConfigSource::kSlotA : ConfigSource::kSlotB,
            ConfigLoadStatus::kOk, true, a, b};
  }

  return load_from_single_valid(a.ok ? a : b, a, b);
}

const char* to_string(ConfigDecodeReason reason) {
  switch (reason) {
    case ConfigDecodeReason::kOk: return "OK";
    case ConfigDecodeReason::kEmpty: return "EMPTY";
    case ConfigDecodeReason::kLength: return "LENGTH";
    case ConfigDecodeReason::kMagic: return "MAGIC";
    case ConfigDecodeReason::kSchema: return "SCHEMA";
    case ConfigDecodeReason::kPayloadLength: return "PAYLOAD_LENGTH";
    case ConfigDecodeReason::kFlags: return "FLAGS";
    case ConfigDecodeReason::kCrc: return "CRC";
    case ConfigDecodeReason::kBounds: return "BOUNDS";
  }
  return "UNKNOWN";
}

const char* to_string(ConfigSource source) {
  switch (source) {
    case ConfigSource::kFactory: return "FACTORY";
    case ConfigSource::kSlotA: return "A";
    case ConfigSource::kSlotB: return "B";
  }
  return "UNKNOWN";
}

const char* to_string(ConfigLoadStatus status) {
  switch (status) {
    case ConfigLoadStatus::kOk: return "OK";
    case ConfigLoadStatus::kFactoryDefaults: return "FACTORY_DEFAULTS";
    case ConfigLoadStatus::kConfigCorrupt: return "CONFIG_CORRUPT";
    case ConfigLoadStatus::kBackupDegraded: return "BACKUP_DEGRADED";
    case ConfigLoadStatus::kRedundant: return "REDUNDANT";
    case ConfigLoadStatus::kGenerationConflict: return "GENERATION_CONFLICT";
    case ConfigLoadStatus::kGenerationAmbiguous: return "GENERATION_AMBIGUOUS";
  }
  return "UNKNOWN";
}

}  // namespace ventilation
